File: 88.merge-sorted-array.cpp
/*
 * @lc app=leetcode id=88 lang=cpp
 *
 * [88] Merge Sorted Array
 */
#include <iostream>
#include <vector>
using namespace std;
// @lc code=start
class Solution {
 public:
  void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {

    while (m > 0 && n > 0) {
      if (nums1[m - 1] > nums2[n - 1]) {
        nums1[m + n - 1] = nums1[m - 1];
        m--;
      } else {
        nums1[m + n - 1] = nums2[n - 1];
        n--;
      }
    }

    if (m == 0) {
      for (int i = 0; i < n; i++) {
        nums1[i] = nums2[i];
      }
    }

  }
};
// @lc code=end

File: Top Interview Questions/src/88.merge-sorted-array.java
/*
 * @lc app=leetcode id=88 lang=java
 *
 * [88] Merge Sorted Array
 */

// @lc code=start
class Solution {
  public void merge(int[] nums1, int m, int[] nums2, int n) {
    if (m == 0) {
      System.arraycopy(nums2, 0, nums1, 0, n);
    }
    
    int cursor = nums1.length - 1;
    m--; n--;
    while (m >= 0 && n >= 0) {
      if (nums1[m] > nums2[n]) {
        nums1[cursor] = nums1[m--];
      } else {
        nums1[cursor] = nums2[n--];
      }
      cursor--;
    }

    if (n >= 0) {
      System.arraycopy(nums2, 0, nums1, 0, n + 1);
    }
  }
}
